packages/driver/src/cy/keyboard.ts

```diff
@@ -0,0 +1,1156 @@
+  return _.some(_.omit(modifiers, ['shift']))
+}
+
+// type shouldBe = <K extends Function,T>(fn:K, a:T): re
+// const shouldBeNumber = (a) => (_.isNumber(a) ? a : throwError())
+// const shouldBe = (fn: (t1)=>t1 is number, a)=>fn(a)?a:throwError()
+
+/**
+ * @example '{foo}' => 'foo'
+ */
+const parseCharsBetweenCurlyBraces = (chars: string) => {
+  return /{(.+?)}/.exec(chars)![1]
+}
+
+const shouldIgnoreEvent = <
+  T extends KeyEventType,
+  K extends { [key in T]?: boolean }
+>(
+  eventName: T,
+  options: K
+) => {
+  return options[eventName] === false
+}
+
+const shouldUpdateValue = (el: HTMLElement, key: KeyDetails) => {
+  if (!key.text) return true
+
+  const bounds = $selection.getSelectionBounds(el)
+  const noneSelected = bounds.start === bounds.end
+
+  if ($elements.isInput(el) || $elements.isTextarea(el)) {
+    if ($elements.isReadOnlyInputOrTextarea(el)) {
+      return false
+    }
+
+    if (noneSelected) {
+      const ml = $elements.getNativeProp(el, 'maxLength')
+
+      //# maxlength is -1 by default when omitted
+      //# but could also be null or undefined :-/
+      //# only care if we are trying to type a key
+      if (ml === 0 || ml > 0) {
+        //# check if we should update the value
+        //# and fire the input event
+        //# as long as we're under maxlength
+        if (!($elements.getNativeProp(el, 'value').length < ml)) {
+          return false
+        }
+      }
+    }
+  }
+
+  return true
+}
+
+const getKeymap = () => {
+  return {
+    ...keyboardMappings,
+    ...modifierChars,
+    // TODO: add the reset of USKeyboard to available keys
+    // ...USKeyboard,
+  }
+}
+const validateTyping = (
+  el: HTMLElement,
+  keys: KeyDetails[],
+  currentIndex: number,
+  onFail: Function,
+  skipCheckUntilIndex?: number,
+) => {
+
+  const chars = joinKeyArrayToString(keys.slice(currentIndex))
+  const allChars = joinKeyArrayToString(keys)
+
+  if (skipCheckUntilIndex) {
+    return { skipCheckUntilIndex: skipCheckUntilIndex-- }
+  }
+
+  debug('validateTyping:', chars, el)
+
+  // debug('validateTyping', el, chars)
+  const $el = $dom.wrap(el)
+  const numElements = $el.length
+  const isBody = $el.is('body')
+  const isTextLike = $dom.isTextLike(el)
+  let isDate = false
+  let isTime = false
+  let isMonth = false
+  let isWeek = false
+  let isDateTime = false
+
+  if ($elements.isInput(el)) {
+    isDate = $dom.isType(el, 'date')
+    isTime = $dom.isType(el, 'time')
+    isMonth = $dom.isType(el, 'month')
+    isWeek = $dom.isType(el, 'week')
+    isDateTime =
+      $dom.isType(el, 'datetime') || $dom.isType(el, 'datetime-local')
+  }
+
+  const isFocusable = $elements.isFocusable($el)
+  const clearChars = '{selectall}{delete}'
+  const isClearChars = _.startsWith(chars.toLowerCase(), clearChars)
+
+  //# TODO: tabindex can't be -1
+  //# TODO: can't be readonly
+
+  if (isBody) {
+    return {}
+  }
+
+  if (!isFocusable && !isTextLike) {
```

Comment:
These 2 if statements are confusing. We're going to throw an error if `!isFocusable`, then we are further distinguishing the error shown by whether `isTextLike` or not. So, maybe:

```suggestion
  if (!isFocusable) {
   const node = $dom.stringify($el)
   const errMsg = isTextLike ? 'not_actionable_textlike' : 'not_on_typeable_element' 

    $utils.throwErrByPath(`type.${errMsg}`, {
      onFail,
      args: { node },
    })
 
  }
```